src/ray/core_worker/transport/normal_task_submitter.cc

```diff
@@ -39,53 +38,43 @@ Status NormalTaskSubmitter::SubmitTask(TaskSpecification task_spec) {
-        auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
-        scheduling_key_entry.task_queue.push_back(task_spec);
-        scheduling_key_entry.resource_spec = task_spec;
-
-        if (!scheduling_key_entry.AllWorkersBusy()) {
-          // There are idle workers, so we don't need more
-          // workers.
-
-          for (auto active_worker_addr : scheduling_key_entry.active_workers) {
-            RAY_CHECK(worker_to_lease_entry_.find(active_worker_addr) !=
-                      worker_to_lease_entry_.end());
-            auto &lease_entry = worker_to_lease_entry_[active_worker_addr];
-            if (!lease_entry.is_busy) {
-              OnWorkerIdle(active_worker_addr,
-                           scheduling_key,
-                           /*was_error*/ false,
-                           /*error_detail*/ "",
-                           /*worker_exiting*/ false,
-                           lease_entry.assigned_resources);
-              break;
-            }
+    absl::MutexLock lock(&mu_);
+    if (cancelled_tasks_.find(task_spec.TaskId()) != cancelled_tasks_.end()) {
+      cancelled_tasks_.erase(task_spec.TaskId());
+    } else {
+      task_spec.GetMutableMessage().set_dependency_resolution_timestamp_ms(
+          current_sys_time_ms());
+      // Note that the dependencies in the task spec are mutated to only contain
+      // plasma dependencies after ResolveDependencies finishes.
+      const SchedulingKey scheduling_key(
+          task_spec.GetSchedulingClass(),
+          task_spec.GetDependencyIds(),
+          task_spec.IsActorCreationTask() ? task_spec.ActorCreationId() : ActorID::Nil(),
+          task_spec.GetRuntimeEnvHash());
+      auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
+      scheduling_key_entry.task_queue.push_back(task_spec);
+      scheduling_key_entry.resource_spec = task_spec;
+
+      if (!scheduling_key_entry.AllWorkersBusy()) {
+        // There are idle workers, so we don't need more
+        // workers.
+
+        for (const auto &active_worker_addr : scheduling_key_entry.active_workers) {
+          RAY_CHECK(worker_to_lease_entry_.find(active_worker_addr) !=
+                    worker_to_lease_entry_.end());
+          auto &lease_entry = worker_to_lease_entry_[active_worker_addr];
+          if (!lease_entry.is_busy) {
+            OnWorkerIdle(active_worker_addr,
+                         scheduling_key,
+                         /*was_error*/ false,
+                         /*error_detail*/ "",
+                         /*worker_exiting*/ false,
+                         lease_entry.assigned_resources);
+            break;
           }
         }
-        RequestNewWorkerIfNeeded(scheduling_key);
       }
-    }
-    if (!keep_executing) {
-      RAY_UNUSED(task_finisher_->FailOrRetryPendingTask(
-          task_spec.TaskId(), rpc::ErrorType::TASK_CANCELLED, nullptr));
+      RequestNewWorkerIfNeeded(scheduling_key);
```

Comment:
we removed a call to FailOrRetryPendingTask here. is this expected?

Reply:
ya this is the call we removed for when the task is removed and the task dependencies have just now been resolved, since now we're failing task before when cancel is actually called